### DISCUSSION TITLE
[IIncrementalGenerator] Multiple files same symbol

### DISCUSSION THREAD

**Discussion by workgroupengineering** [dotnet/roslyn#71575] [Q&A] [answered]
## Structure

considering following project structure:

```
MyProject 
│
├ MyNamespace ┬ MyObject.cs
│             ├ MyObject.Do.cs
│             ├ MyObject.Properties.cs 
┋             ┋
```

`MyObject.*.cs`  is same symbol, but it is split into more files for better readability.

## Use case

I'm trying to create an incremental generator that generates a code when the symbol has an attribute that implements a specific interface.
To do this, I created `CreateSyntaxProvider` which filters all `ClassDeclarationSyntax` nodes and in the transformation function obtains the symbol via ` syntaxContext.SemanticModel.GetDeclaredSymbol(syntaxContext.Node, token)` and checks that it has an attribute that implements a specific interface

## The Issue

From the tests I have done, I see that the generator is invoked for each file that makes up the symbol, creating duplicate code.
How can I avoid this problem?

**Comment by CyrusNajmabadi** [comment, marked as answer]
You'll need your incremental generator pipeline to .Collect all the elements and process them in aggregate to know what to merge and what to filter or throw out.

Note:

> I'm trying to create an incremental generator that generates a code when the symbol has an attribute

You should use .ForAttributeWithMetadataName for this purpose.  It's normally around 100x faster and uses around 100x memory to accomplish the same task.  As a bonus, it is symbol (not syntax) based.  So you'd only get this symbol a single time, even though it has multiple parts spread across multiple files.
